### PR TITLE
ARTEMIS-1895 - Add duplicate metadata failure callback to ActiveMQServerPlugin

### PR DESCRIPTION
**Pull request by cshannon**
Add a callback on duplicate metadata which will allow extra functionality to be added.

### CODE REVIEW

File: artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java [hunk 1]
```diff
@@ -263,6 +264,8 @@ private void validateClientID(ClientSession validateSession, String clientID) th
       } catch (ActiveMQException e) {
          if (e.getType() == ActiveMQExceptionType.DUPLICATE_METADATA) {
             throw new InvalidClientIDException("clientID=" + clientID + " was already set into another connection");
+         } else {
+            throw new InvalidClientIDException("Error setting clientID=" + clientID + ": " + e.getMessage());
```
Comment (michaelandrepearce) [inline artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQConnection.java]: The original exception here may not be related to InvalidClientId, seems a bit dangerous to catchall and throw like this. Maybe bubble up or convert the exception

**Comment by cshannon** [comment]
@michaelandrepearce - good point, i fixed it to bubble up the ActiveMQException and so it is handled properly by the client

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/SessionMetadataAddExceptionTest.java [hunk 2]
```diff
@@ -75,6 +88,24 @@ public void testInvalidClientIdSetFactory() throws Exception {
...
+   @Test(timeout = 5000, expected = JMSException.class)
+   public void testDuplicateClientIdSet() throws Exception {
+      ActiveMQConnectionFactory activeMQConnectionFactory = (ActiveMQConnectionFactory) cf;
+      Connection con = cf.createConnection();
+      Connection con2 = cf.createConnection();
+      try {
+         con.setClientID("valid");
+         con2.setClientID("valid");
+         fail("Should have failed for duplicate clientId");
+      } catch (Exception e) {
```
Comment (michaelandrepearce) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/SessionMetadataAddExceptionTest.java]: Should this test that explicit exception type is thrown not just any old exception. As probably should fail if an unexpectes exception is ever thrown.

Comment (cshannon) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/SessionMetadataAddExceptionTest.java]: I switched it to catch the InvalidClientIdException that gets thrown

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/SessionMetadataAddExceptionTest.java [hunk 3]
```diff
@@ -75,6 +93,37 @@ public void testInvalidClientIdSetFactory() throws Exception {
...
+      ActiveMQConnectionFactory activeMQConnectionFactory = (ActiveMQConnectionFactory) cf;
+      Connection con = cf.createConnection();
+      Connection con2 = cf.createConnection();
+      try {
+         con.setClientID("valid");
+         con2.setClientID("valid");
+         fail("Should have failed for duplicate clientId");
+      } catch (InvalidClientIDException e) {
+         assertEquals(1, duplicateCount.get());
+         throw e;
```
Comment (michaelandrepearce) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/SessionMetadataAddExceptionTest.java]: Problem here is assert will be bypassed and test will still pass if another exception thrown as tests expects JMSException.

Expects should be explicit and need a catch all with and assert fail

**Comment by cshannon** [comment]
I removed the re-throw and the expect for the JMSException so it will only pass on InvalidClientIDException as that is the only type that should be thrown

**Comment by michaelandrepearce** [comment]
Looks good to me

**Comment by cshannon** [comment]
@clebertsuconic - is this ok to merge?

**Comment by cshannon** [comment]
I don't think it's related as the test it failed on doesn't have a plugin registered.  I tried to see if I could trigger it to re-run the build in travis ci but i didn't see a way to do that.

**Comment by clebertsuconic** [comment]
The easiest way to rebuild is to. 

Got commit —amend
Got push -f

**Comment by cshannon** [comment]
The build looks good now

**Comment by clebertsuconic** [comment]
thanks a lot.. merging

**Pull request closed without merging**